aws/s3.go

```diff
@@ -0,0 +1,295 @@
+	result, err := svc.GetBucketLocation(input)
+	if err != nil {
+		return "", err
+	}
+
+	if result.LocationConstraint == nil {
+		// GetBucketLocation returns nil for us-east-1
+		// https://github.com/aws/aws-sdk-go/issues/1687
+		return "us-east-1", nil
+	}
+	return *result.LocationConstraint, nil
+}
+
+// getS3BucketTags returns S3 Bucket tags.
+func getS3BucketTags(svc *s3.S3, bucketName *string) ([]map[string]string, error) {
+	input := &s3.GetBucketTaggingInput{
+		Bucket: bucketName,
+	}
+
+	tags := []map[string]string{}
+
+	result, err := svc.GetBucketTagging(input)
+	if err != nil {
+		return tags, err
+	}
+
+	for _, tagSet := range result.TagSet {
+		tags = append(tags, map[string]string{"Key": *tagSet.Key, "Value": *tagSet.Value})
+	}
+
+	return tags, err
+}
+
+// getAllS3Buckets lists and returns a map of per region AWS S3 buckets which were created before excludeAfter
+func getAllS3Buckets(session *session.Session, excludeAfter time.Time, bucketNameSubStr string) (map[string][]*string, error) {
+	svc := s3.New(session)
+
+	input := &s3.ListBucketsInput{}
+
+	output, err := svc.ListBuckets(input)
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	var bucketNamesPerRegion = make(map[string][]*string)
+
+OUTER:
+	for _, bucket := range output.Buckets {
+		logging.Logger.Debugf("Checking - Bucket %s", *bucket.Name)
+
+		if len(bucketNameSubStr) > 0 {
+			if !strings.Contains(*bucket.Name, bucketNameSubStr) {
+				logging.Logger.Debugf("Skipping - Bucket %s - failed substring filter - %s", *bucket.Name, bucketNameSubStr)
+				continue
+			}
+		}
+
+		if !excludeAfter.After(*bucket.CreationDate) {
+			logging.Logger.Debugf("Skipping - Bucket %s - matched CreationDate filter", *bucket.Name)
+			continue
+		}
+
+		bucketTags, err := getS3BucketTags(svc, bucket.Name)
+		if len(bucketTags) > 0 {
+			for _, tagSet := range bucketTags {
+				if tagSet["Key"] == "cloud-nuke-excluded" && tagSet["Value"] == "true" {
+					logging.Logger.Infof("Skipping - Bucket %s - matched tag filter", *bucket.Name)
+					continue OUTER
+				}
+			}
+		}
+
+		bucketRegion, err := getS3BucketRegion(svc, bucket.Name)
+		if err != nil {
+			logging.Logger.Warnf("Skipping - Bucket %s - Failed to get bucket location.", *bucket.Name)
+			continue
+		}
+
+		if _, ok := bucketNamesPerRegion[bucketRegion]; !ok {
+			bucketNamesPerRegion[bucketRegion] = []*string{}
+		}
+
+		bucketNamesPerRegion[bucketRegion] = append(bucketNamesPerRegion[bucketRegion], bucket.Name)
+	}
+
+	return bucketNamesPerRegion, nil
+}
+
+func getS3BucketObjectVersions(svc *s3.S3, bucketName *string, objectKey *string) ([]*string, error) {
+	versions := []*string{}
+
+	err := svc.ListObjectVersionsPages(
+		&s3.ListObjectVersionsInput{
+			Bucket: bucketName,
+			Prefix: objectKey,
+		},
+		func(page *s3.ListObjectVersionsOutput, lastPage bool) (shouldContinue bool) {
+			for _, obj := range page.Versions {
+				versions = append(versions, obj.VersionId)
+			}
+			return true
+		},
```

Comment:
You also need to get the `DeleteMarkers` and handle those.

Reply:
I might be missing something but I did not follow this - DeleteMarker is set when you do a simple Delete on a versioned bucket. But in case of a versioned bucket we are deleting each object with version so that it gets permanently deleted. https://docs.amazonaws.cn/en_us/AmazonS3/latest/dev/DeletingObjectVersions.html - because we are doing https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.DeleteObjects will we run into this?

Reply:
Checked and understood. This is to handle the state when someone does a simple delete and the deletemarker is created. Will add the support.

Reply:
Done